### PR TITLE
Adding scafacos example

### PR DESCRIPTION
**Pull request by junghans**
/cc @rhalver

### CODE REVIEW

File: core/example/scafacos_example.cpp [hunk 1]
```diff
@@ -0,0 +1,290 @@
+#include <Cabana_Types.hpp>
+#include <Cabana_AoSoA.hpp>
+#include <Cabana_MemberDataTypes.hpp>
```
Comment (junghans) [inline core/example/scafacos_example.cpp]: @sslattery what is the new header to use here?

Comment (sslattery) [reply core/example/scafacos_example.cpp]: `Cabana_MemberTypes.hpp`

**Comment by junghans** [comment]
Now it is at:
```
[ 12%] Building CXX object core/example/CMakeFiles/ScafacosExample.dir/scafacos_example.cpp.o
/home/junghans/computing/Cabana/core/example/scafacos_example.cpp:45:13: error: ‘MemberDataTypes’ in namespace ‘Cabana’ does not name a template type
     Cabana::MemberDataTypes<float,                        // (0) x-position type
             ^~~~~~~~~~~~~~~
```

**Comment by junghans** [comment]
```
[ 12%] Building CXX object core/example/CMakeFiles/ScafacosExample.dir/scafacos_example.cpp.o
/home/junghans/computing/Cabana/core/example/scafacos_example.cpp:59:29: error: ‘InnerArrayLayout’ in namespace ‘Cabana’ does not name a template type
 using ArrayLayout = Cabana::InnerArrayLayout<32,Cabana::LayoutRight>;
                             ^~~~~~~~~~~~~~~~
```
@sslattery what is the new name for `InnerArrayLayout`?

File: core/example/CMakeLists.txt [hunk 2]
```diff
@@ -3,5 +3,12 @@ if(NOT Cabana_ENABLE_EXAMPLES)
   return()
 endif()
 
+find_package(PkgConfig)
+pkg_check_modules(SCAFACOS scafacos IMPORTED_TARGET)
```
Comment (dalg24) [inline core/example/CMakeLists.txt]: Why did you chose this over `find_package(scafacos)`?

File: core/example/scafacos_example.cpp [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+
+// Designate the types that the particles will hold.
+using ParticleDataTypes =
+    Cabana::MemberTypes<float,                        // (0) x-position type
+                        float,                        // (1) y-position type
+                        float,                        // (2) z-position type
+                        double[space_dim],            // (3) velocity type
+                        double,			              // (4) charge
+		                double,			              // (5) potential
+		                double[space_dim],		      // (6) electric field values
```
Comment (dalg24) [inline core/example/scafacos_example.cpp]: Did you mean to indent tis way?

File: core/example/scafacos_example.cpp [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+    auto p_x = particles.slice<PositionX>();
+    auto p_y = particles.slice<PositionY>();
+    auto p_z = particles.slice<PositionZ>();
+    auto v = particles.slice<Velocity>();
+    auto q = particles.slice<Charge>();
+    auto pot = particles.slice<Potential>();
+    auto field = particles.slice<Field>();
+    auto indx = particles.slice<Index>();
+
+    for ( auto idx = 0; idx != particles.size(); ++idx )
```
Comment (dalg24) [inline core/example/scafacos_example.cpp]: `auto` will deduce `int` and you will get warning about comparing signed and unsigned integers.

File: core/example/scafacos_example.cpp [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+            field( idx, d ) = 0.0;
+
+        // Create alternating charge
+        q(idx) = ((idx_x + idx_y + idx_z)%2)?1.0:-1.0;
+
+        // Set potential
+        pot(idx) = 0.0;
+
+        // Set global particle index
+        indx(idx) = idx+1;
```
Comment (dalg24) [inline core/example/scafacos_example.cpp]: Any particular reason why you chose to index from 1?

File: core/example/scafacos_example.cpp [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+    {
+        // Calculate location of particle in crystal
+        int idx_x = idx % crystal_size;
+        int idx_y = (idx / crystal_size) % crystal_size;
+        int idx_z = idx / (crystal_size * crystal_size);	
+	
+        // Initialize position.
+        p_x(idx) = (double)idx_x * 1.0;
+        p_y(idx) = (double)idx_y * 1.0;
+        p_z(idx) = (double)idx_z * 1.0;
```
Comment (dalg24) [inline core/example/scafacos_example.cpp]: This feels unnecessary, just do `p_z(idx) = idx;`

File: core/example/scafacos_example.cpp [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+const int VectorLength = 8;
+
+// Set the type for the particle AoSoA.
+using ParticleList = Cabana::AoSoA<ParticleDataTypes,MemorySpace,VectorLength>;
+
+//---------------------------------------------------------------------------//
+// Helper functions.
+//---------------------------------------------------------------------------//
+// Function to intitialize the particles.
+void initializeParticles( ParticleList particles, int crystal_size )
```
Comment (dalg24) [inline core/example/scafacos_example.cpp]: I know it's only 48 bytes, but I would still pass the particles by reference.

File: core/example/scafacos_example.cpp [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+    auto pot = particles.slice<Potential>();
+    auto field = particles.slice<Field>();
+    auto indx = particles.slice<Index>();
+
+    for ( auto idx = 0; idx < particles.size(); ++idx )
+    {
+        std::cout << "Position: "
+                  << p_x(idx) << " "
+                  << p_y(idx) << " "
+                  << p_z(idx) << std::endl;
```
Comment (dalg24) [inline core/example/scafacos_example.cpp]: s/std::end/'\n'/

File: core/example/scafacos_example.cpp [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+      MPI_Abort(MPI_COMM_WORLD, 1);
+    else
+      std::cout << "WARNING: Continuing after error!" << std::endl;
+    return false;
+  }
+  return true;
+}
+
+// example main
+void exampleMain(int num_particle, int crystal_size, std::string method, int mpi_comm, int rank)
```
[the remark this answers is not in the document]
Comment (dalg24) [reply core/example/scafacos_example.cpp]: Also probably should pass the string by const reference.

File: core/example/CMakeLists.txt (hunk 2 above)
Comment (dalg24) [reply core/example/CMakeLists.txt]: Also did you consider having an option `Cabana_ENABLE_SCAFACOS` that defaults to false and adding the `REQUIRED` argument when you look for it.

Comment (junghans) [reply core/example/CMakeLists.txt]: I am usually more a fan of "enable everything you can automatically". CMake has the `CMAKE_DISABLE_FIND_PACKAGE_<FOO>=ON` options to disable stuff.

**Comment by junghans** [comment]
@rhalver I added scafacos to the CI, but there is a linking error:
```
[ 72%] [32m[1mLinking CXX executable ScafacosExample[0m
cd /home/travis/build/junghans/Cabana/build/core/example && /usr/local/cmake-3.9.2/bin/cmake -E cmake_link_script CMakeFiles/ScafacosExample.dir/link.txt --verbose=1
/usr/lib/ccache/g++   -march=native -O3 -DNDEBUG   CMakeFiles/ScafacosExample.dir/scafacos_example.cpp.o  -o ScafacosExample ../src/libcabanacore.a /home/travis/kokkos/lib/libkokkos.a -ldl -lhwloc -fopenmp /home/travis/scafacos/lib/libfcs4fortran.a /home/travis/scafacos/lib/libfcs.a /home/travis/scafacos/lib/libfcs_direct.a /home/travis/scafacos/lib/libfcs_ewald.a /home/travis/scafacos/lib/libfcs_fmm.a /home/travis/scafacos/lib/libfcs_memd.a /home/travis/scafacos/lib/libfcs_mmm1d.a /home/travis/scafacos/lib/libfcs_mmm2d.a /home/travis/scafacos/lib/libfcs_p2nfft.a /usr/lib/libgsl.so /usr/lib/libgslcblas.so /home/travis/scafacos/lib/libfcs_p3m.a /home/travis/scafacos/lib/libfcs_pepc.a /home/travis/scafacos/lib/libfcs_pp3mg.a /home/travis/scafacos/lib/libfcs_vmg.a /home/travis/scafacos/lib/libfcs_wolf.a /home/travis/scafacos/lib/libfcs_near.a /home/travis/scafacos/lib/libfcs_gridsort.a /home/travis/scafacos/lib/libfcs_redist.a /home/travis/scafacos/lib/libfcs_resort.a /home/travis/scafacos/lib/libfcs_common.a /home/travis/scafacos/lib/libfcs_mmm.a /home/travis/scafacos/lib/libfcs_pnfft.a /home/travis/scafacos/lib/libfcs_pfft.a /home/travis/scafacos/lib/libfcs_fftw3_mpi.a /home/travis/scafacos/lib/libfcs_fftw3.a /usr/lib/libmpi_f90.so /usr/lib/libmpi_f77.so /usr/lib/gcc/x86_64-linux-gnu/4.8/libgfortran.so /usr/lib/gcc/x86_64-linux-gnu/4.8/libgomp.so /usr/lib/gcc/x86_64-linux-gnu/4.8/libquadmath.so /usr/lib/libmpi_cxx.so /usr/lib/libmpi.so /usr/lib/x86_64-linux-gnu/libdl.so /usr/lib/x86_64-linux-gnu/libhwloc.so /usr/lib/gcc/x86_64-linux-gnu/4.8/libstdc++.so /usr/lib/x86_64-linux-gnu/libm.so /usr/lib/gcc/x86_64-linux-gnu/4.8/libgcc_s.so -lpthread 
/home/travis/scafacos/lib/libfcs_vmg.a(libfcs_vmg_la-com_check_residual.o): In function `VMG::CommandProxy<VMGCommandCheckResidual>::~CommandProxy()':
/home/travis/build/junghans/Cabana/scafacos-1.0.1/lib/vmg/src/./base/proxy.hpp:49: undefined reference to `VMG::CommandProxyBase::~CommandProxyBase()'
/home/travis/scafacos/lib/libfcs_vmg.a(libfcs_vmg_la-com_check_residual.o): In function `~CommandProxy':
/home/travis/build/junghans/Cabana/scafacos-1.0.1/lib/vmg/src/./base/proxy.hpp:49: undefined reference to `VMG::CommandProxyBase::~CommandProxyBase()'
/home/travis/scafacos/lib/libfcs_vmg.a(libfcs_vmg_la-com_check_residual.o):(.rodata._ZTIN3VMG12CommandProxyI23VMGCommandCheckResidualEE[_ZTIN3VMG12CommandProxyI23VMGCommandCheckResidualEE]+0x10): undefined reference to `typeinfo for VMG::CommandProxyBase'
```
seems like some symbols couldn't be resolved.

**Comment by junghans** [comment]
Linking issue resolved.

**Comment by rhalver** [comment]
I included the remarks, i.e. replaced the std::endl with '\n' and used references instead of passing the
construct by value. The indention should be corrected and the indices of the particles start with zero now (left over of the past Fortran developer days, where we started to count the particles with 1). The example was extended in order to be able to run in parallel, so that the crystal is now divided between the different processes with a simple domain decomposition and computation can take place in parallel, before the crystal was duplicated on each process, which of course led to wrong results due to overlapping charges.

**Comment by junghans** [comment]
Ping @sslattery

File: core/example/scafacos_example.cpp [hunk 10]
```diff
@@ -158,7 +158,7 @@ void printParticles( const ParticleList particles, parallel_info& info )
               << '\n'
               << '\n';
 
-    for ( auto idx = 0; idx < particles.size(); ++idx )
+    for ( int idx = 0; idx < particles.size(); ++idx )
```
[the remark this answers is not in the document]
Comment (rhalver) [reply core/example/scafacos_example.cpp]: I got confused and replace both in the first commit. Afterwards I changed it to unsigned, but of course size_t would be the better solution, sorry for the clutter.

File: core/example/CMakeLists.txt (hunk 2 above)
Comment (dalg24) [reply core/example/CMakeLists.txt]: > Why did you chose this over find_package(scafacos)?

You didn't address that comment

File: core/example/scafacos_example.cpp [hunk 11]
```diff
@@ -0,0 +1,444 @@
...
+#include <Cabana_AoSoA.hpp>
+#include <Cabana_MemberTypes.hpp>
+
+#include <Kokkos_Core.hpp>
+
+#include <cstdlib>
+#include <iostream>
+#include <string>
+
+#include "fcs.h"
```
Comment (dalg24) [inline core/example/scafacos_example.cpp]: I don't remember any coding guidelines for Cabana but this is a bit inconsistent. I would expect you to use angle brackets and place it above the includes from the standard library.

File: core/example/scafacos_example.cpp [hunk 12]
```diff
@@ -0,0 +1,444 @@
...
+    auto q = particles.slice<Charge>();
+    auto pot = particles.slice<Potential>();
+    auto field = particles.slice<Field>();
+    auto indx = particles.slice<Index>();
+
+    int offset;
+
+    //compute offset with MPI_Exscan
+    MPI_Exscan(&(info.n_local_particles), &offset, 1, MPI_INT, MPI_SUM, info.comm);
+    if (info.rank == 0) offset = 0;
```
Comment (dalg24) [inline core/example/scafacos_example.cpp]: Maybe add a comment here that the MPI standard says that the receive buffer on rank 0 is undefined.

File: core/example/CMakeLists.txt (hunk 2 above)
Comment (junghans) [reply core/example/CMakeLists.txt]: I was just too lazy to write a FindScafacos.cmake module.

Comment (junghans) [reply core/example/CMakeLists.txt]: Not sure, I don't see Scafacos providing any CMake files!

Comment (junghans) [reply core/example/CMakeLists.txt]: I got it, v1.0.1 doesn't have these CMake files, but we can change the once scafacos 1.1 got released.

Comment (junghans) [reply core/example/CMakeLists.txt]: @mh0fm4, when is the next release of scafacos planned?

File: core/example/scafacos_example.cpp (hunk 3 above)
Comment (junghans) [reply core/example/scafacos_example.cpp]: Fixed!

**Comment by junghans** [comment]
@sslattery, this is ready to merge!

**Comment by junghans** [comment]
ping @sslattery

**Pull request merged**